File: using-typescript/factory-pattern.ts
interface Button {
    render(): void;
    onClick(f: Function): void;
}

class WindowsButton implements Button {
    public render(): void {
        console.info("This is a Windows Button")
    }

    public onClick(f: Function): void {
        console.info("Window button click functionality")
    }
}

class HTMLButton implements Button {
    public render(): void {
        console.info("This is an HTML Button")
    }

    public onClick(f.Function): void {
        console.info("HTML button click functionality")
    }
}

abstract class Dialog {
    abstract createButton(): Button;

    closeDialog(): void {
        // Implement for closing dialog
    }

    render(): void {
        const okButton = this.createButton();
        okButton.onClick(this.closeDialog);
        okButton.render();
    }
}

class WindowsDialog implements Dialog {
    public createButton(): Button {
        return new WindowsButton();
    }
}

class HTMLDialog implements Dialog {
    public createButton(): Button {
        return new HTMLButton();
    }
}

class Application {
    private dialog: Dialog | null = null;

    readApplicationConfigFile(): {Target:string} {
        // Dummy method to read application config file
        return { Target = 'Windows'};
    }

    initialize(): void {
        const config = this.readApplicationConfigFile();
        if (config.Target == "Windows") {
            this.dialog = new WindowsDialog();
        } else if (config.Target == "Web") {
            this.dialog = new HTMLButton();
        } else {
            throw new Error("Error! Unknown target environment")
        }
    }

    main(): void {
        this.initialize();
        if (this.dialog) {
            this.dialog.render();
        }
    }
}

// Usage
const app = new Application();
app.main();

File: using-python/factory_pattern.py
from abc import ABC, abstractmethod

# Product interface
class Button(ABC):
    
    @abstractmethod
    def render(self):
        pass

    @abstractmethod
    def on_click(self, f):
        pass


# Concrete products
class WindowsButton(Button):
    def render(self):
        # Render a button in Windows style
        print("This is a Windows Button")
        pass

    def on_click(self, f):
        # Bind a native OS click event
        pass


class HTMLButton(Button):
    def render(self):
        # Render an HTML representation of a button
        print("This is an HTML Button")
        pass

    def on_click(self, f):
        # Bind a web browser click event
        pass

# Creator class
class Dialog(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        pass

    def close_dialog(self):
        pass

    def render(self):
        ok_button = self.create_button()
        ok_button.on_click(self.close_dialog)
        ok_button.render()


# Concrete Creators
class WindowsDialog(Dialog):
    def create_button(self) -> Button:
        return WindowsButton()
    

class WebDialog(Dialog):
    def create_button(self) -> Button:
        return HTMLButton()

# Application class
class Application:

    def __init__(self) -> None:
        self.dialog = None

    def read_application_config_file(self):
        # Dummy method to read application config file
        return {'Target': 'Windows'}

    def initialize(self):
        config = self.read_application_config_file()
        if config['Target'] == "Windows":
            self.dialog = WindowsDialog()
        elif config['Target'] == "Web":
            self.dialog = WebDialog()
        else:
            raise Exception("Error! Unknown environment passed.")
        
    def main(self):
        self.initialize()
        self.dialog.render()


# Usage
app = Application()
app.main()